text/0000-immovable-types.md

```diff
@@ -0,0 +1,144 @@
+
+We can easily work around issues with this in code by using a single borrow of the immovable value and just reborrow.
+
+Illegal:
+```rust
+let mutex = StaticMutex::new(0);
+*mutex.lock() += 1;
+*mutex.lock() += 1;
+```
+Workaround using reborrows:
+```rust
+let mutex = &StaticMutex::new(0);
+*mutex.lock() += 1;
+*mutex.lock() += 1;
+```
+
+A borrow such as `&var.field` where `field` is immovable will last as long as the lifetime of `var` to ensure it matches the lifetime of the field.
+
+We need to prevent assignment to immovable types once they have been borrowed. This is because assignment actually moves the l-value before calling `Drop::drop`. If there are any restrictions on the l-value or if the l-value has a dereference operation, assignment to immovable types is not allowed.
+
+Types which implement `Copy`, but not `Move` are allowed. You can still copy them around, but borrows follows the restrictions of `?Move` types.
+
+## Immovable types contained in movable types
+
+To allow immovable types to be contained in movable types, we introduce a `core::cell::MobileCell` wrapper which itself implements `Move`. It works similarly to `Cell` in that it disallows references to the value inside.
+```rust
+#[lang = "mobile_cell"]
+pub struct MobileCell<T: ?Move> {
+	value: T,
+}
+
+impl<T: ?Move> MobileCell<T> {
+	pub const fn new(value: T) -> Movable<T> {
+		Movable {
+			value: value,
+		}
+	}
+
+	pub fn into_inner(self) -> T {
+		self.value
+	}
+
+	pub fn replace(&mut self, new_value: T) -> T {
+		let mut result = MobileCell::new(new_value);
+		core::mem::replace(self, &mut result);
+		result.into_inner()
+	}
+}
+```
+
+## Implications for language traits
+
+In order to allow functions to take immovable types and arguments and return them, we need to change `FnOnce`, `FnMut` and `Fn`. A `?Move` bound should be added for the `Args` type parameter to these traits. We also need a `?Move` bound on `FnOnce::Output`, which is backwards incompatible. `FnOnce::Output` was stabilized in 1.12, so hopefully there aren't any code relying on it yet.
+
+Having a `?Move` bound on `Deref::Target` would be nice. It would allow us to use the dereference operator on `Box`, `Rc`, and `Arc` containing immovable types.
+
+A `?Move` bound on `IntoIterator::IntoIter` and `Iterator::Self` would also be useful, since you could then use immovable iterators in for-loops.
+
+I suggest we do a `crater` run to investigate if these breakages are feasible.
+
+Changing these associated types will be insta-stable. You would be unable to write stable code which would conflict with this proposal. `?Move` bounds would also show up in documentation, although we would be able to filter those out if desired.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
```

Comment:
*Documentation RFC pedantry hat on here.* 🤓

This section could use to be expanded a bit. Specifically, if accepted this RFC would need to include updates:

- to the standard library where the types are implemented
- to the Reference, under [9. Special Traits](https://doc.rust-lang.org/reference.html#special-traits)
- (probably) to the book in the FFI section

I also think the hand-wave that "the concept is likely familiar to users of C, C++, and C FFIs" is insufficient. This is something which people writing FFIs for the first time will need a place to learn, and we can't assume that everyone writing an FFI has prior C or C++ experience. There are many, and an increasing number of, Rust users who are coming to Rust wanting to do FFI for e.g. Python or Ruby and who haven't done it before in other contexts.

That said, discussions of the immovability in the C, C++, and FFI contexts might be *really* helpful prior art—it could possibly be cited, and it should certainly be considered as useful background for whoever writes the docs if this is accepted!